### PR TITLE
Add data_format option to convolution2d_transpose

### PR DESCRIPTION
**Pull request by thuyen**
Add `data_format` option to `convolution2d_transpose`.

### CODE REVIEW

**Comment by thuyen** [comment]
@zhangyaobit Could you have a look? :)

**Review by zhangyaobit — APPROVED** [review]
Looks good overall! 

Please update the comments for "inputs: a tensor of size [batch_size, height, width, channels]." at https://github.com/tensorflow/tensorflow/pull/4986/files#diff-94bbcef0ec8a5cdef55f705e99c2b2edR914

File: tensorflow/contrib/layers/python/layers/layers.py [hunk 1]
```diff
@@ -942,14 +944,23 @@ def convolution2d_transpose(
...
   with variable_scope.variable_scope(
       scope, 'Conv2d_transpose', [inputs], reuse=reuse) as sc:
+    if data_format not in (DATA_FORMAT_NCHW, DATA_FORMAT_NHWC):
+      raise ValueError('data_format has to be either NCHW or NHWC.')
     dtype = inputs.dtype.base_dtype
     kernel_h, kernel_w = utils.two_element_tuple(kernel_size)
     stride_h, stride_w = utils.two_element_tuple(stride)
-    num_filters_in = utils.last_dimension(
-        inputs.get_shape(), min_rank=4)
+    if data_format==DATA_FORMAT_NCHW:
```
Comment (zhangyaobit) [inline tensorflow/contrib/layers/python/layers/layers.py]: Spaces around ==

**Comment by thuyen** [comment]
@zhangyaobit Just updated the request with your suggestions.

**Review by zhangyaobit — COMMENTED** [review]
Could you add a test for convolution2d_transpose for the output shape when using NCHW,  and a test for conv2d_transpose for NCHW in kernel_tests/conv2d_transpose_test.py?

**Comment by thuyen** [comment]
@zhangyaobit Tests are added and have passed locally. One weird thing for `kernel_tests/conv2d_transpose_test.py` is that if I keep `tf.nn.conv2d_transpose` inside `self.test_session` (like the way tests are done for `NHWC` data format), `tensorflow` will call the `CPU` op instead of `GPU` op. Could you explain why?

File: tensorflow/python/kernel_tests/conv2d_transpose_test.py [hunk 2]
```diff
@@ -157,6 +158,129 @@ def testGradient(self):
     err_tolerance = 0.0005
     self.assertLess(err, err_tolerance)
 
+  def get_available_gpus(self):
+    local_device_protos = device_lib.list_local_devices()
+    return [x.name for x in local_device_protos if x.device_type == 'GPU']
+
+  def testConv2DTransposeSingleStrideNCHW(self):
+    # `NCHW` data fomat is only supported for `GPU` device.
+    if len(self.get_available_gpus()) < 1:
```
Comment (zhangyaobit) [inline tensorflow/python/kernel_tests/conv2d_transpose_test.py]: Could you use is_gpu_available() and do something like this? https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/bias_op_test.py#L92

Note that you can use self.test_session(use_gpu=use_gpu) to put the op on GPU.

Comment (zhangyaobit) [inline tensorflow/python/kernel_tests/conv2d_transpose_test.py]: Yeah, your observation is expected. test_session by default put the op on CPU; you can use self.test_session(use_gpu=True) to change this default setting.

Comment (thuyen) [inline tensorflow/python/kernel_tests/conv2d_transpose_test.py]: Thanks Yao! Just updated the pull with your suggestions.

**Comment by zhangyaobit** [comment]
Jenkins, test this please.

**Comment by zhangyaobit** [comment]
Linux CPU Tests failed. NCHW is not supported yet for conv2d_transpose :(

https://ci.tensorflow.org/job/tensorflow-pull-requests-cpu/2217/console

**Comment by zhangyaobit** [comment]
GPU NCHW support seems to be there; you may just need to add is_gpu_available() in your test.

**Comment by thuyen** [comment]
Yeah, that error is in CPU op. Just narrow the `layers_test` for `conv2d_transpose` on GPU only for `NCHW` data format.

**Comment by zhangyaobit** [comment]
Jenkins, test this please.

**Comment by thuyen** [comment]
Can you test this again @zhangyaobit. Sorry I missed one test in previous commit.

**Comment by zhangyaobit** [comment]
Jenkins, test this please.

**Comment by zhangyaobit** [comment]
Jenkins, test this please.

**Comment by zhangyaobit** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
Fails on Mac with this:
FAIL: testEarlyDominatesBootstrap (__main__.FinishedNodesTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/var/tmp/_bazel_jenkins/13e370a18c169b19baeafefb05212b85/execroot/tensorflow-pull-requests-mac/bazel-out/local-opt/bin/tensorflow/contrib/tensor_forest/finished_nodes_op_test.runfiles/org_tensorflow/tensorflow/contrib/tensor_forest/python/kernel_tests/finished_nodes_op_test.py", line 141, in testEarlyDominatesBootstrap
    self.assertAllEqual([4], finished.eval())
  File "/private/var/tmp/_bazel_jenkins/13e370a18c169b19baeafefb05212b85/execroot/tensorflow-pull-requests-mac/bazel-out/local-opt/bin/tensorflow/contrib/tensor_forest/finished_nodes_op_test.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 482, in assertAllEqual
    "Shape mismatch: expected %s, got %s." % (a.shape, b.shape))
AssertionError: Shape mismatch: expected (1,), got (2,).


Could you look into this?

**Comment by thuyen** [comment]
It seems unrelated to this pull request I guess. We didn't touch anything that could possibly affect tensor_forest.

**Comment by thuyen** [comment]
All tests have passed now. Do you know why I one failed on Mac before?

**Pull request merged**